src/controllerclientimpl.h

```diff
@@ -64,6 +64,9 @@ class ControllerClientImpl : public ControllerClient, public boost::enable_share
     virtual void DeleteDirectoryOnController_UTF8(const std::string& desturi);
     virtual void DeleteDirectoryOnController_UTF16(const std::wstring& desturi);
 
+    virtual void SaveBackup(std::vector<unsigned char>& vdata, bool config, bool media, double timeout);
+    virtual void RestoreBackup(const std::vector<unsigned char>& vdata, bool config, bool media);
```

Comment:
```
virtual void SaveBackup(std::ostream& outputStream, bool config, bool media, double timeout);
virtual void RestoreBackup(const std::istream& inputStream, bool config, bool media);
```

User could open file as stream, or use in memory buffer

```
std::ofstream ofs("backup.tar", std::ofstream::out);
```

Reply:
I have written the following script:

```
#!/usr/bin/python
import socket
sock = socket.socket()
sock.connect(('127.0.0.1',80))
if 1:
    sock.send('''POST /fileupload HTTP/1.1
Host: localhost
X-CSRFToken: csrftoken
Accept: */*
User-Agent: upload_test
Connection: close
Referer: http://localhost
Cookie: csrftoken=csrftoken
Content-Type: multipart/form-data; boundary=section
Authorization: Basic bXVqaW46bXVqaW4=
Transfer-Encoding: chunked

5b
--section
Content-Disposition: form-data; name="file"; filename="file"

abcdefghijklmn
1b
opqrstuvwxyz
--section--
0
'''.replace('\n','\r\n'))
else:
    sock.send('''POST /fileupload HTTP/1.1
Host: localhost
X-CSRFToken: csrftoken
Accept: */*
User-Agent: upload_test
Connection: close
Referer: http://localhost
Cookie: csrftoken=csrftoken
Content-Type: multipart/form-data; boundary=section
Authorization: Basic bXVqaW46bXVqaW4=
Content-Length: 118

--section
Content-Disposition: form-data; name="file"; filename="file"

abcdefghijklmn
opqrstuvwxyz
--section--
'''.replace('\n','\r\n'))
print sock.recv(4096)
```

When I disable chunk, I can upload the file. However when I enable chuck, firstly webstack returned the following error:

```
411 Length Required
A request of the requested method POST requires a valid Content-length.
```

stackoverflow says I can put `WSGIChunkedRequest On`, so I retried. Then I got internal server error with the error:

```
Request data read error when proxying data to daemon process: Unknown error -102.
```

Now I need to say that I need to precompute Content-Length (i.e. CURLFORM_CONTENTSLENGTH). So istream cannot be used and ifstream needs to be used.
For controllerclientpy, https://github.com/psf/requests/blob/master/requests/models.py#L159 indicates that they are reading whole file to compute Content-Length.
For ostream I don't have such concerns yet.

Reply:
What's the unknown error -102? Any log message  in webstack.log?

Reply:
Also, istream doesn't have seekg and tellg?

Reply:
> unknown error -102

that was all webstack.log printed.

> istream doesn't have seekg and tellg

sure, istream does support them. the caveat is (in `controllerclient->RestoreBackup(cin)`) when cin is '< input.tar', seek will success, but when cin is `cat input.tar |`, seek will fail. I added error handling and document.

So, restoring concern is now cleared, right?

Reply:
Right, but you didn't fix `SaveBackup`, `std::ostream& outputStream` they should be symmetric